src/Functions/FunctionVariantAdaptor.cpp

```diff
@@ -0,0 +1,746 @@
+    std::unordered_map<ColumnVariant::Discriminator, size_t> variant_indexes;
+    const auto & local_discriminators = variant_column.getLocalDiscriminators();
+    const auto & offsets = variant_column.getOffsets();
+
+    for (size_t i = 0; i != local_discriminators.size(); ++i)
+    {
+        auto local_discr = local_discriminators[i];
+        if (local_discr == ColumnVariant::NULL_DISCRIMINATOR)
+        {
+            selector.push_back(0);
+            variants_offsets.emplace_back();
+        }
+        else
+        {
+            auto global_discr = variant_column.globalDiscriminatorByLocal(local_discr);
+            /// Check if we already allocated selector index for this variant type.
+            /// If not, append it to list of variants and remember its index.
+            auto it = variant_indexes.find(global_discr);
+            if (it == variant_indexes.end())
+            {
+                it = variant_indexes.emplace(global_discr, variants.size()).first;
+                variants.emplace_back(variant_column.getVariantPtrByLocalDiscriminator(local_discr), variant_types[global_discr], "");
+            }
+
+            selector.push_back(it->second);
+            variants_offsets.push_back(offsets[i]);
+        }
+    }
+
+    /// Create set of arguments for each variant using selector.
+    std::vector<ColumnsWithTypeAndName> variants_arguments;
+    variants_arguments.resize(variants.size());
+    for (size_t i = 0; i != arguments.size(); ++i)
+    {
+        if (i == variant_argument_index)
+        {
+            for (size_t j = 1; j != variants_arguments.size(); ++j)
+                variants_arguments[j].push_back(variants[j]);
+        }
+        else
+        {
+            auto columns = arguments[i].column->scatter(variants.size(), selector);
+            for (size_t j = 0; j != variants_arguments.size(); ++j)
+                variants_arguments[j].emplace_back(std::move(columns[j]), arguments[i].type, arguments[i].name);
+        }
+    }
+
+    /// Execute function over all created sets of arguments and remember all results.
+    std::vector<ColumnPtr> variants_results;
+    std::vector<DataTypePtr> variants_result_types;
+    variants_results.reserve(variants.size());
+    variants_result_types.reserve(variants.size());
+    /// 0 index is allocated for rows with NULL values, it doesn't have any result,
+    /// we will insert NULL values in these rows.
+    variants_results.emplace_back();
+    variants_result_types.emplace_back();
+
+    for (size_t i = 1; i != variants_arguments.size(); ++i)
+    {
+        try
+        {
+            auto func_base = function_overload_resolver->build(variants_arguments[i]);
+            auto nested_result_type = func_base->getResultType();
+            auto nested_result = func_base->execute(variants_arguments[i], nested_result_type, variants_arguments[i][0].column->size(), dry_run)
+                                     ->convertToFullColumnIfConst();
```

Comment:
Should we cover `func_base->execute` in `try/catch` as well? I am not sure about it, if something is wrong during function execution with compatible arguments, we should throw it instead of just ignoring the result